### PR TITLE
[serve] Fix traceback string for RayTaskErrors when deploying serve app

### LINKED ISSUE #33118: [Serve] Syntax error when deploying serve app will show as `no detail available` upon fetching serve status
### What happened + What you expected to happen

If a syntax error occurs when Serve tries to import a Serve application and deploy it, the error will not show up correctly on `serve status`. Specifically it just shows `ray.exceptions.RayTaskError(SyntaxError): <no detail available>`. This makes it hard for users to debug.

### Versions / Dependencies

ray 2.3.0

### Reproduction script

Source file:
```
# fail.py
x = (1 + 2 # syntax error

from ray import serve
@serve.deployment
...
```
Config:
```
# config.yaml
host: 127.0.0.1
port: 8000
import_path: fail:app
```

Deploy with `serve deploy config.yaml` -> `serve status` gives the output:
```
name: ''
app_status:
  status: DEPLOY_FAILED
  message: |
    Deployment failed:
    Traceback (most recent call last):
      File "/Users/cindyz/ray/python/ray/serve/_private/application_state.py", line 162, in update
        ray.get(finished[0])
      File "/Users/cindyz/ray/python/ray/_private/client_mode_hook.py", line 105, in wrapper
        return func(*args, **kwargs)
      File "/Users/cindyz/ray/python/ray/_private/worker.py", line 2384, in get
        raise value.as_instanceof_cause()
    ray.exceptions.RayTaskError(SyntaxError): <no detail available>
  deployment_timestamp: 1678228862.220114
deployment_statuses: []
```

### Issue Severity

None

### PR DESCRIPTION
**Pull request by zcin**
## Why are these changes needed?

If a syntax error occurs when Serve tries to import a Serve application and deploy it, the error will not show up correctly on `serve status`. Specifically it just shows `ray.exceptions.RayTaskError(SyntaxError): <no detail available>`. This makes it hard for users to debug. 

This is because for some reason `traceback.format_exc()` does not get the traceback string correctly for `RayTaskError`s; it seems to have it's own way of getting the traceback string [here](https://github.com/ray-project/ray/blob/a892241ca7574af47f278a667e6493a4b03686d7/python/ray/exceptions.py#L163). So we need to change `str(e)`. 

Previously:
```
name: ''
app_status:
  status: DEPLOY_FAILED
  message: |
    Deployment failed:
    Traceback (most recent call last):
      File "/Users/cindyz/ray/python/ray/serve/_private/application_state.py", line 162, in update
        ray.get(finished[0])
      File "/Users/cindyz/ray/python/ray/_private/client_mode_hook.py", line 105, in wrapper
        return func(*args, **kwargs)
      File "/Users/cindyz/ray/python/ray/_private/worker.py", line 2384, in get
        raise value.as_instanceof_cause()
    ray.exceptions.RayTaskError(SyntaxError): <no detail available>
  deployment_timestamp: 1678228862.220114
deployment_statuses: []
```

With `str(e)`:
```
name: ''
app_status:
  status: DEPLOY_FAILED
  message: |-
    Deployment failed:
    ray::run_graph() (pid=81000, ip=10.104.98.7)
      File "/Users/cindyz/ray/python/ray/serve/controller.py", line 859, in run_graph
        graph = import_attr(import_path)
      File "/Users/cindyz/miniforge3/envs/ray/lib/python3.8/importlib/__init__.py", line 127, in import_module
        return _bootstrap._gcd_import(name[level:], package, level)
      File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
      File "<frozen importlib._bootstrap>", line 991, in _find_and_load
      File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
      File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
      File "<frozen importlib._bootstrap_external>", line 839, in exec_module
      File "<frozen importlib._bootstrap_external>", line 976, in get_code
      File "<frozen importlib._bootstrap_external>", line 906, in source_to_code
      File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
      File "/Users/cindyz/Desktop/fail.py", line 4
        print("yo")
        ^
    SyntaxError: invalid syntax
  deployment_timestamp: 1678228961.0310981
deployment_statuses: []
```

## Related issue number

Closes #33118

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by zcin** [comment]
Does anyone know how to get the lint checker to ignore a file?

**Comment by edoakes** [comment]
@zcin you can add it to the excludes list here:
https://github.com/ray-project/ray/blob/35ffe75f6a6acbbbafc87aa52e178e6c34ac7db7/.flake8#L1

File: python/ray/serve/_private/application_state.py [hunk 1]
```diff
@@ -160,9 +160,9 @@ def update(self):
                     return
                 try:
                     ray.get(finished[0])
-                except RayTaskError:
+                except RayTaskError as e:
                     self.status = ApplicationStatus.DEPLOY_FAILED
-                    self.app_msg = f"Deployment failed:\n{traceback.format_exc()}"
+                    self.app_msg = f"Deployment failed:\n{str(e)}"
```
Comment (edoakes) [inline python/ray/serve/_private/application_state.py:165]: how does this change the formatting & why is it required?

Comment (zcin) [reply python/ray/serve/_private/application_state.py:165]: I added an example of how it changes the formatting in the PR description. I haven't dug into why yet (only saw this [custom formatting function](https://github.com/ray-project/ray/blob/a892241ca7574af47f278a667e6493a4b03686d7/python/ray/exceptions.py#L163)), I can look into it more later today.

Comment (edoakes) [reply python/ray/serve/_private/application_state.py:165]: oh -- thanks for doing that :)

that's fine, let's leave a comment indicating that it is important and shouldn't be changed to `traceback.format_exc()` (which is more standard in my experience)

Comment (zcin) [reply python/ray/serve/_private/application_state.py:165]: Sg, will do!

**Review by edoakes — APPROVED** [review]

**Review by sihanwang41 — APPROVED** [review]
thank you! nice fix

**Comment by zcin** [comment]
I'm not sure if it's possible to get linter to ignore a file with syntax error, since it can't even parse the file :(

**Comment by sihanwang41** [comment]
> I'm not sure if it's possible to get linter to ignore a file with syntax error, since it can't even parse the file :(

Can you give a try? https://docs.pylint.org/faq.html#how-can-i-tell-pylint-to-never-check-a-given-module

**Comment by edoakes** [comment]
@zcin looks like the linter error is now coming from `black`, I believe you can pass `--exclude` or `--extend_exclude` wherever we run the `black` command:
https://black.readthedocs.io/en/stable/usage_and_configuration/file_collection_and_discovery.html

**Comment by edoakes** [comment]
This looks like the place:
https://github.com/ray-project/ray/blob/7bb5d648b01e8809bd16ee9ff2e1ec1a172fa054/ci/lint/format.sh#L151

**Comment by ollie-iterators** [comment]
Something that could work if this doesn't is pylint: disable-next = ...

File: ci/lint/format.sh [hunk 2]
```diff
@@ -149,11 +149,13 @@ MYPY_FILES=(
...
-    '--force-exclude' 'python/ray/core/src/ray/gcs/*'
-    '--force-exclude' 'python/ray/thirdparty_files/*'
-    '--force-exclude' 'python/ray/_private/thirdparty/*'
+    '--force-exclude'
+    'python/ray/cloudpickle/*|'`
+    `'python/build/*|'`
+    `'python/ray/core/src/ray/gcs/*|'`
+    `'python/ray/thirdparty_files/*|'`
+    `'python/ray/_private/thirdparty/*|'`
+    `'python/ray/serve/tests/test_config_files/syntax_error\.py'
```
Comment (zcin) [inline ci/lint/format.sh:158]: I'm pretty sure `--force-exclude` only takes one regex string, so all the files need to be in that single regex string. When I tried adding `python/ray/serve/tests/test_config_files/syntax_error.py` as a `--force-exclude` [here](https://buildkite.com/ray-project/oss-ci-build-pr/builds/14136#0186c7a2-b24c-4a4c-9ca7-b4dc14560334), the linter ignored `syntax_error.py` but started reformatting a bunch of files in `python/ray/_private/thirdparty/*`, so the last `--force-exclude` probably overrides all the previous ones.

**Review by shrekris-anyscale — APPROVED** [review]
Nice work!

**Pull request merged**